### ISSUE TITLE
meson --wipe forgets about CFLAGS and friends

### ISSUE THREAD

**Issue by dcbaker** [mesonbuild/meson#4650]
Someone in the mesa community noticed that if you configure with CFLAGS and friends, then call meson --wipe that CFLAGS and friends are forgotten. Can be demonstrated with something like

```meson
project('test', 'cpp', default_options : ['cpp_std=c++03'])
executable('main', 'main.cpp')
```
```c++
#include <iostream>
#include <vector>
#include <string>

int main () {
  std::vector<std::string> v{ "one", "two", "three" };
  for (auto const & s: v) {
    std::cout << s << std::endl;
  }
}
```
```sh
CXXFLAGS="-std=c++11" meson build
ninja -C build
meson --wipe build
ninja -C build
```

Obviously this is ugly since it relies on setting the c++ std through the environment, but it's a pretty easy way to ensure a failure.

**Comment by xclaesse** [comment]
Yes, that's by design, we remember only the command line, the use-case is really just like in autotools where you replay the configure command you find in config.log.

I think a way to fix this is the suggestion to generate a native file instead.

**Comment by xclaesse** [comment]
Personally I think that if you rely on environment, then it's your duty to always have it setup. In the few projects I'm working on where env matters, we have a script that setup env and you cannot do anything without sourcing that bash script first. Must set PATH for a few tools, etc.

**Comment by dcbaker** [comment]
Except that we promise that once we read CFLAGS and friends they are forever. `ninja clean` doesn't erase them, but if ninja decides that it needs to invoke `meson --wipe` will silently reconfigure your build options is absolutely awful and unfriendly.

I guess extending and generating a native-file is my next big project.

**Comment by RoyiAvital** [comment]
There is also another small issue with `--wipe`.
In case the build folder is already empty it triggers an error and stop the build process.

I think it should either trigger only a warning or just in case the build folder is empty proceed without any message.
Certainly no reason to break the whole process.

**Comment by denizzzka** [comment]
Looks like this issue is leads to things like https://github.com/mesonbuild/meson/issues/7831 (option value default setting is ignored)

**Comment by nanthony21** [comment]
I've also noticed that `--wipe` removes the convenient `.gitignore` file from the build directory.

**Comment by xclaesse** [comment]
> I've also noticed that `--wipe` removes the convenient `.gitignore` file from the build directory.

I did not know meson generate .gitignore, fixed this in https://github.com/mesonbuild/meson/pull/8670.

**Comment by xclaesse** [comment]
> Except that we promise that once we read CFLAGS and friends they are forever. `ninja clean` doesn't erase them, but if ninja decides that it needs to invoke `meson --wipe` will (nearly) silently change your build options is absolutely awful and unfriendly.

Note that ninja will never use --wipe. It's meant to be used by the user to be sure to get a pristine build from scratch. It is also the only way to reconfigure with a new env. I really think this is not a bug but a very useful feature.

**Comment by bonzini** [comment]
See also #4637.
